File: contents/flood_fill/code/cpp/flood_fill.cpp
#include <array>
#include <cassert>
#include <iostream>
#include <queue>
#include <stack>
#include <vector>

using CartesianIndex = std::array<int, 2>;

auto inbounds(CartesianIndex size, CartesianIndex loc) {
  if (loc[0] < 0 || loc[1] < 0) {
    return false;
  } else if (loc[0] >= size[0] || loc[1] >= size[1]) {
    return false;
  }
  return true;
}

auto find_neighbors(
    std::vector<std::vector<float>> const& grid,
    CartesianIndex loc,
    float old_value,
    float /* new_value */) {

  const std::vector<CartesianIndex> possible_neighbors{
      {loc[0], loc[1] + 1},
      {loc[0] + 1, loc[1]},
      {loc[0], loc[1] - 1},
      {loc[0] - 1, loc[1]}};

  std::vector<CartesianIndex> neighbors;

  for (auto const& possible_neighbor : possible_neighbors) {
    const auto size = CartesianIndex{
        static_cast<int>(grid[0].size()), static_cast<int>(grid.size())};
    const auto x = static_cast<std::size_t>(possible_neighbor[0]);
    const auto y = static_cast<std::size_t>(possible_neighbor[1]);
    if (inbounds(size, possible_neighbor) && grid[x][y] == old_value) {
      neighbors.push_back(possible_neighbor);
    }
  }

  return neighbors;
}

void recursive_fill(
    std::vector<std::vector<float>>& grid,
    CartesianIndex loc,
    float old_value,
    float new_value) {
  if (old_value == new_value) {
    return;
  }

  const auto x = static_cast<std::size_t>(loc[0]);
  const auto y = static_cast<std::size_t>(loc[1]);

  grid[x][y] = new_value;

  const auto possible_neighbors = find_neighbors(grid, loc, old_value, new_value);
  for (auto const& possible_neighbor : possible_neighbors) {
    recursive_fill(grid, possible_neighbor, old_value, new_value);
  }
}

void queue_fill(
    std::vector<std::vector<float>>& grid,
    CartesianIndex loc,
    float old_value,
    float new_value) {
  if (old_value == new_value) {
    return;
  }

  auto q = std::queue<CartesianIndex>{};
  q.push(loc);
  const auto x = static_cast<std::size_t>(loc[0]);
  const auto y = static_cast<std::size_t>(loc[1]);
  grid[x][y] = new_value;

  while (q.size() > 0) {
    const auto current_loc = q.front();
    q.pop();
    const auto possible_neighbors =
        find_neighbors(grid, current_loc, old_value, new_value);
    for (auto const& neighbor : possible_neighbors) {
      const auto neighbor_x = static_cast<std::size_t>(neighbor[0]);
      const auto neighbor_y = static_cast<std::size_t>(neighbor[1]);
      grid[neighbor_x][neighbor_y] = new_value;
      q.push(neighbor);
    }
  }
}

void stack_fill(
    std::vector<std::vector<float>>& grid,
    CartesianIndex loc,
    float old_value,
    float new_value) {
  if (old_value == new_value) {
    return;
  }

  auto s = std::stack<CartesianIndex>{};
  s.push(loc);

  while (s.size() > 0) {
    const auto current_loc = s.top();
    s.pop();

    const auto x = static_cast<std::size_t>(current_loc[0]);
    const auto y = static_cast<std::size_t>(current_loc[1]);

    if (grid[x][y] == old_value) {
      grid[x][y] = new_value;
      const auto possible_neighbors =
          find_neighbors(grid, current_loc, old_value, new_value);
      for (auto const& neighbor : possible_neighbors) {
        s.push(neighbor);
      }
    }
  }
}

int main() {

  const std::vector<std::vector<float>> grid{
      {0, 0, 1, 0, 0},
      {0, 0, 1, 0, 0},
      {0, 0, 1, 0, 0},
      {0, 0, 1, 0, 0},
      {0, 0, 1, 0, 0}};

  const std::vector<std::vector<float>> solution_grid{
      {1, 1, 1, 0, 0},
      {1, 1, 1, 0, 0},
      {1, 1, 1, 0, 0},
      {1, 1, 1, 0, 0},
      {1, 1, 1, 0, 0}};

  const CartesianIndex start_loc{1, 1};

  auto test_grid = grid;
  recursive_fill(test_grid, start_loc, 0.0, 1.0);
  assert(test_grid == solution_grid);

  test_grid = grid;
  queue_fill(test_grid, start_loc, 0.0, 1.0);
  assert(test_grid == solution_grid);

  test_grid = grid;
  stack_fill(test_grid, start_loc, 0.0, 1.0);
  assert(test_grid == solution_grid);

  return EXIT_SUCCESS;
}

File: contents/flood_fill/code/csharp/FloodFill.cs
using System.Linq;
using System.Diagnostics;
using System.Collections.Generic;


namespace Graphics
{
    abstract class FloodFill
    {
        // A simple point on the 2 dimensional integer plane.
        private class Point2I
        {
            // Coordinates
            public int x;
            public int y;

            public Point2I(int x, int y)
            {
                this.x = x;
                this.y = y;
            }
        }

        // Utility object for comparing List<List<float>> objects.
        private class FloatListEqualityComparer : IEqualityComparer<List<float>>
        {
            public bool Equals(List<float> one, List<float> two)
            {
                return ReferenceEquals(one, two) || one != null && two != null && one.SequenceEqual(two);
            }

            public int GetHashCode(List<float> list)
            {
                return list.GetHashCode();
            }
        }

        private static bool InBounds(Point2I size, Point2I loc)
        {
            return loc.x >= 0 && loc.y >= 0 && loc.x < size.x && loc.y < size.y;
        }

        private static List<Point2I> FindNeighbors(ref List<List<float>> grid, Point2I loc, float oldValue)
        {
            var possibleNeighbors = new List<Point2I>
            {
                new Point2I(loc.x, loc.y + 1),
                new Point2I(loc.x + 1, loc.y),
                new Point2I(loc.x, loc.y - 1),
                new Point2I(loc.x - 1, loc.y)
            };
            var neighbors = new List<Point2I>();
            var size = new Point2I(grid[0].Count, grid.Count);

            foreach (Point2I possibleNeighbor in possibleNeighbors)
            {
                var x = possibleNeighbor.x;
                var y = possibleNeighbor.y;
                if (!InBounds(size, possibleNeighbor)) {
                    continue;
                }
                if (grid[x][y].Equals(oldValue)) {
                    neighbors.Add(possibleNeighbor);
                }
            }
            return neighbors;
        }

        private static void RecursiveFill(ref List<List<float>> grid, Point2I loc, float oldValue, float newValue)
        {
            if (oldValue.Equals(newValue)) {
                return;
            }
            grid[loc.x][loc.y] = newValue;

            var possibleNeighbors = FindNeighbors(ref grid, loc, oldValue);
            foreach (Point2I possibleNeighbor in possibleNeighbors) {
                RecursiveFill(ref grid, possibleNeighbor, oldValue, newValue);
            }
        }

        private static void QueueFill(ref List<List<float>> grid, Point2I loc, float oldValue, float newValue)
        {
            if (oldValue.Equals(newValue)) {
                return;
            }

            var queue = new Queue<Point2I>();
            queue.Enqueue(loc);
            grid[loc.x][loc.y] = newValue;

            while (queue.Count > 0)
            {
                var currentLoc = queue.Dequeue();
                var possibleNeighbors = FindNeighbors(ref grid, currentLoc, oldValue);
                foreach (Point2I neighbor in possibleNeighbors)
                {
                    grid[neighbor.x][neighbor.y] = newValue;
                    queue.Enqueue(neighbor);
                }
            }
        }

        private static void StackFill(ref List<List<float>> grid, Point2I loc, float oldValue, float newValue)
        {
            if (oldValue.Equals(newValue)) {
                return;
            }

            var stack = new Stack<Point2I>();
            stack.Push(loc);

            while (stack.Count > 0)
            {
                var currentLoc = stack.Pop();

                var x = currentLoc.x;
                var y = currentLoc.y;

                if (grid[x][y].Equals(oldValue))
                {
                    grid[x][y] = newValue;
                    var possibleNeighbors = FindNeighbors(ref grid, currentLoc, oldValue);
                    foreach (Point2I neighbor in possibleNeighbors) {
                        stack.Push(neighbor);
                    }
                }
            }
        }


        private static void Main(string[] args)
        {
            // All neighbouring zeros, adjacent to (1, 1), must be replaced with 3 in the end result.
            var grid = new List<List<float>>
            {
                new List<float>(){0, 0, 1, 0, 0},
                new List<float>(){0, 0, 1, 0, 0},
                new List<float>(){0, 0, 1, 0, 0},
                new List<float>(){8, 0, 1, 1, 1},
                new List<float>(){0, 0, 0, 0, 0}
            };
            var solutionGrid = new List<List<float>>
            {
                new List<float>(){3, 3, 1, 0, 0},
                new List<float>(){3, 3, 1, 0, 0},
                new List<float>(){3, 3, 1, 0, 0},
                new List<float>(){8, 3, 1, 1, 1},
                new List<float>(){3, 3, 3, 3, 3}
            };
            var startingPoint = new Point2I(1, 1);
            var gridComparator = new FloatListEqualityComparer();

            var testGrid = new List<List<float>>(grid);
            RecursiveFill(ref testGrid, startingPoint, 0, 3);
            Debug.Assert(testGrid.SequenceEqual(solutionGrid, gridComparator), "Recursive Flood Fill Failed");

            testGrid = new List<List<float>>(grid);
            QueueFill(ref testGrid, startingPoint, 0, 3);
            Debug.Assert(testGrid.SequenceEqual(solutionGrid, gridComparator), "Queue Flood Fill Failed");

            testGrid = new List<List<float>>(grid);
            StackFill(ref testGrid, startingPoint, 0, 3);
            Debug.Assert(testGrid.SequenceEqual(solutionGrid, gridComparator), "Stack Flood Fill Failed");

        }
    }
}
